src/backend/drm/compositor/mod.rs

```diff
@@ -519,42 +523,90 @@ impl<B: AsRef<framebuffer::Handle>> FrameState<B> {
         res
     }
 
-    fn commit(&self, surface: &DrmSurface, event: bool) -> Result<(), crate::backend::drm::error::Error> {
+    fn commit(
+        &mut self,
+        surface: &DrmSurface,
+        supports_fencing: bool,
+        event: bool,
+    ) -> Result<(), crate::backend::drm::error::Error> {
         surface.commit(
             self.planes
-                .iter()
+                .iter_mut()
                 // Filter out any skipped planes
                 .filter(|(_, state)| !state.skip)
                 .map(|(handle, state)| super::surface::PlaneState {
                     handle: *handle,
-                    config: state.config.as_ref().map(|config| super::PlaneConfig {
-                        src: config.src,
-                        dst: config.dst,
-                        alpha: config.alpha,
-                        transform: config.transform,
-                        damage_clips: config.damage_clips.as_ref().map(|d| d.blob()),
-                        fb: *config.buffer.as_ref(),
+                    config: state.config.as_mut().map(|config| {
+                        // Try to extract a native fence out of the supplied sync point if any
+                        // If the sync point has no native fence or the surface does not support
+                        // fencing force a wait
+                        if let Some((sync, fence)) = config.sync.as_mut() {
+                            if supports_fencing && fence.is_none() {
+                                *fence = sync.native().map(Arc::new);
+                            }
+
+                            if fence.is_none() {
+                                sync.wait();
```

Comment:
Do we want to wait between each map or accumulate each fence that needs to be waited on and wait in a single spot?

Reply:
I changed that part a bit so that it now also applies to `test_state`. There is a new `build_planes` function that tries to extract the native fence and wait if that fails (if we are not only testing the state).